go/tools/builders/nogo_change.go

```
@@ -0,0 +1,350 @@
+		}
+		size += len(edit.New) + edit.Start - edit.End
+		lastEnd = edit.End
+	}
+
+	return edits, size, nil
+}
+
+
+// NewChangeFromDiagnostics builds a Change from a set of diagnostics.
+// Unlike Diagnostic, Change is independent of the FileSet given it uses perf-file offsets instead of token.Pos.
+// This allows Change to be used in contexts where the FileSet is not available, e.g., it remains applicable after it is saved to disk and loaded back.
+// See https://github.com/golang/tools/blob/master/go/analysis/diagnostic.go for details.
+func NewChangeFromDiagnostics(entries []diagnosticEntry, fileSet *token.FileSet) (*Change, error) {
+	c := NewChange()
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		return c, fmt.Errorf("Error getting current working directory: (%v)", err)
+	}
+
+	var allErrors []error
+
+	for _, entry := range entries {
+		analyzer := entry.Analyzer.Name
+		for _, sf := range entry.Diagnostic.SuggestedFixes {
+			for _, edit := range sf.TextEdits {
+				start, end := edit.Pos, edit.End
+				if !end.IsValid() {
+					end = start
+				}
+
+				file := fileSet.File(edit.Pos)
+				if file == nil {
+					allErrors = append(allErrors, fmt.Errorf("invalid fix: missing file info for pos %v", edit.Pos))
+					continue
+				}
+				if start > end {
+					allErrors = append(allErrors, fmt.Errorf("invalid fix: pos %v > end %v", start, end))
+					continue
+				}
+				if eof := token.Pos(file.Base() + file.Size()); end > eof {
+					allErrors = append(allErrors, fmt.Errorf("invalid fix: end %v past end of file %v", end, eof))
+					continue
+				}
+
+				edit := Edit{Start: file.Offset(start), End: file.Offset(end), New: string(edit.NewText)}
+				fileRelativePath, err := filepath.Rel(cwd, file.Name())
+				if err != nil {
+					fileRelativePath = file.Name() // fallback logic
+				}
+				c.AddEdit(fileRelativePath, analyzer, edit)
+			}
+		}
+	}
+
+	if len(allErrors) > 0 {
+		return c, fmt.Errorf("errors: %v", allErrors)
```

The errors will be appended to other errors and printed out one by one, one per line. We should either return a list of errors here so they will be appended properly, or format them in the same way as other errors and return one error

---

prefer to return a single error, since all these are in one logical group.

formatted same way as in nogo_main.go